test/basics.jl

```diff
@@ -0,0 +1,45 @@
+string_types = [ASCIIStr, LatinStr, UTF8Str, UCS2Str, UTF16Str, UniStr]
+
+@testset "constructors" begin
+    for strtype in string_types
+        @test convert(strtype, [0x61,0x62,0x63,0x21]) == "abc!"
+        @test convert(strtype, "abc!") == "abc!"
+
+        @test isempty(string())
+        @test eltype(GenericString) == Char
+        @test start("abc") == 1
+        @test cmp("ab","abc") == -1
+        @test "abc" === "abc"
+        @test "ab"  !== "abc"
+        @test string("ab", 'c') === "abc"
+        @test string() === ""
+        codegen_egal_of_strings(x, y) = (x===y, x!==y)
+        @test codegen_egal_of_strings(string("ab", 'c'), "abc") === (true, false)
+        let strs = ["", "a", "a b c", "до свидания"]
+            for x in strs, y in strs
+                @test (x === y) == (object_id(x) == object_id(y))
+            end
+        end
+    end
+end
+
+@testset "{starts,ends}with" begin
```

Comment:
You should probably have separate lists of strings to run the tests on, and divide them into the following categories:
1) only contains ASCII characters
2) only contains ASCII and Latin1 characters
3) only contains BMP characters (i.e. <= 0xffff)
4) contains things like Emojis
For ASCIIStr, you can only use the first list
For LatinStr, you should use the first 2 lists
For UCS2Str, you should use the first 3 lists
For all others (UTF8Str, UTF16Str, UTF32Str), use all 4 lists